### PR TITLE
roachtest/tests: introduce metamorphic testing to cdc

### PR DESCRIPTION
**Pull request by wenyihu6**
Prior to this commit, roachtest/cdc relies solely on periodic checks of
changefeed status and latency. This patch takes the first step to introduce a
metamorphic testing framework.

Given the absence of a way to evaluate the output file correctness yet, this new
approach involves running two changefeeds with different configurations,
retrieving their roachtests’ output files, and comparing their data outputs.

Due to potential duplicates in the changefeed output, the test follows these
steps:
1. create two empty tables with the same scheme as the workload tables 
2. convert parquet data to datums 
3. execute `UPSERT` statements on the tables with the datums to eliminate
duplicates 
4. confirm the identical content of the two tables by checking their
fingerprints

Limitations with this approach include: 
- This solution only works for parquet files as of now. (A round trip conversion
is guaranteed between parquet data format and datums. Other data formats are
more complicated.) 
- INSERT is the only operation involved. 
- Due to the large file size, the test randomly selects one target table for
changefeeds. 
- Currently, the changefeeds use the same configurations. However, we plan to
change this soon following a discussion to determine the specfic configurations
that will be randomized.

Part of: cockroachdb#111066

Release note: None

### CODE REVIEW

**Comment by wenyihu6** [comment]
I ran the test with more tables manually which also passed. This PR randomly selects and checks one table only due to significant test time.

**Comment by wenyihu6** [comment]
We should be able to check operations other than INSERT. But it involves copying content from the tpcc tables since we can't execute DELETE on empty tables. We can add this in later.

**Review by miretskiy — DISMISSED** [review]
> Reviewed 6 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114504)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/ccl/changefeedccl/parquet.go` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOeWAb6iQ23sD1E87i:-NjOeWAb6iQ23sD1E87j:b-p6eeys) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/ccl/changefeedccl/parquet.go#L35)):*
> ```Go
> // column metadata for parquet files in tests.
> var includeParquestTestMetadata = buildutil.CrdbTestBuild ||
> 	envutil.EnvOrDefaultBool("COCKROACH_CHANGEFEED_TESTING_INCLUDE_PARQUET_TEST_METADATA",
> ```

do we need to change the definition here?
Based on https://github.com/cockroachdb/cockroach/pull/111949, roachtests should be compiled with
crdb_test tag now...
___
*[`pkg/ccl/changefeedccl/parquet.go` line 352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjPGoZN33w2wNQdvMXj:-NjPGoZN33w2wNQdvMXk:blylmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/ccl/changefeedccl/parquet.go#L352)):*
> ```Go
> 		return -1, errors.Errorf("could not find column names in parquet metadata")
> 	}
> 	_, columnNameSet, err := deserializeMap(columnsNamesString)
> ```

This is probably rather inefficient to keep regenerating maps on (potentially) multiple calls to this method.
I understand though, why you had to add this method here ; the desire is to expose metadata added by
cdc parquet code (here).

May I suggest instead, to create a subpackage here, let's say called `cdcparquet` (to differentiate it from util/parquet package).
Let's move all of the test metadata related code to that package (addParquetTestMetadata, etc)..
Let's also wrap things a bit so that there is a struct, called `ColumnMetadata` containing 2 maps (key columns/all columns)...

Basically, have a way to extract `ColumnMetadata` from `parquet.ReadDatumsMetadata`.
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjPJDqF-CCbLJpV59hH:-NjPJDqF-CCbLJpV59hI:bgixgwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/cmd/roachtest/tests/cdc.go#L1125)):*
> ```Go
> 			firstFeed.waitForCompletion()
> 
> 			secFeed := ct.newChangefeed(feedArgs{
> ```

let's add a todo to support other formats (cause otherwise, it's not much of metamorphic test).
___
*[`pkg/cmd/roachtest/tests/cdc_helper.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOSMtw1oRsTmmbv-fr:-NjOSMtw1oRsTmmbv-fs:bbp4pob) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/cmd/roachtest/tests/cdc_helper.go#L98)):*
> ```Go
> // see a file for a table other than selectedTargetTable. It is just passed in
> // for a validation check.
> func fetchFilesOfTargetTable(
> ```

nit: perhaps a better name might be `listFilesForTargetTable`?
___
*[`pkg/cmd/roachtest/tests/cdc_helper.go` line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOSkZyDTh68fA5t3UX:-NjOSkZyDTh68fA5t3UY:b-5fmcpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/cmd/roachtest/tests/cdc_helper.go#L122)):*
> ```Go
> 	ctx context.Context, es cloud.ExternalStorage, fileName string,
> ) (downloadedFileName string, _ error) {
> 	reader, _, err := es.ReadFile(context.Background(), fileName, cloud.ReadOptions{NoFileSize: true})
> ```

we probably want to use ctx here, instead of background.
___
*[`pkg/cmd/roachtest/tests/cdc_helper.go` line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOUeOz0u0HGKWoY_QU:-NjOUeOz0u0HGKWoY_QV:b9d6evm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/cmd/roachtest/tests/cdc_helper.go#L159)):*
> ```Go
> // processTable reads the local file specified by fileName, parse the data
> // content, execute upsert statement for the file content into the provided
> // targetTable. Note that only INSERT operations are permitted. If any steps
> ```

nit: it's a bit hard to see why INSERT is mentioned here.. maybe it doesn't belong as
top level function comment.
___
*[`pkg/util/parquet/writer.go` line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOPH0c6bvVR2tUCzDp:-NjOPH0d5dANWIVJ124Q:bok0dya) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/util/parquet/writer.go#L184)):*
> ```Go
> 	// Add additional metadata required to use the reader utility functions in
> 	// testutils.go.
> 	if includeParquestReaderMetadata {
> ```

I would prefer not to add the env variable to a (low-ish) level parquet library.
I think we already propagate the fact that we need to include parquet metadata

Also, I think that we already add metadata from the changefeedccl/parquet.go (the caller of this library) , so I think the use of buildutil.CrdbTestBuild here is purely for this package unit tests.  I think that can stay.
___
*[`pkg/cmd/roachtest/tests/cdc_helper_test.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjORzv8Bcj35N8FNNDp:-NjORzv8Bcj35N8FNNDq:b-56ixhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/cmd/roachtest/tests/cdc_helper_test.go#L38)):*
> ```Go
> }
> 
> // TestUpsertStmtForTable verifies that upsertStmtForTable correctly format
> ```

nit: s/format/formats/

**Review by miretskiy — COMMENTED** [review]
I haven't finished review -- just dumping initial set of comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-:-NjPJSIJ9LPiHybWdaxV:b-7k49f0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

**Review by DarrylWong — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114504)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/parquet.go` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOeWAb6iQ23sD1E87i:-NjSj6IDEpxUHUh92MkN:b-29dvlv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/ccl/changefeedccl/parquet.go#L35)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do we need to change the definition here?
Based on https://github.com/cockroachdb/cockroach/pull/111949, roachtests should be compiled with
crdb_test tag now...
</blockquote></details>

I didn't take a look at anything else, just saw this and figured I'd chime in.

Roachtests will have a 50% chance to use the cockroach short binary w/ crdb_test flag and a 50% chance to use the standard cockroach without.

Meaning if you want this constant to be on all the time in a certain roachtest, you'll have to do something like above. Alternatively, if you don't mind other metamorphic constants being on, you can add `CockroachBinary: registry.RuntimeAssertionsCockroach,` to the test spec. This will force it to always use the binary with crdb_test.

Sidenote: the benchmark test spec will force this test to always use the non crdb_test binary, unless you add the test spec mentioned above

**Comment by wenyihu6** [comment]
*[`pkg/ccl/changefeedccl/parquet.go` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOeWAb6iQ23sD1E87i:-Njh6Sit7YJThxx5mb8k:b-ofsaza) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/ccl/changefeedccl/parquet.go#L35)):*
<details><summary><i>Previously, DarrylWong wrote…</i></summary><blockquote>

I didn't take a look at anything else, just saw this and figured I'd chime in.

Roachtests will have a 50% chance to use the cockroach short binary w/ crdb_test flag and a 50% chance to use the standard cockroach without.

Meaning if you want this constant to be on all the time in a certain roachtest, you'll have to do something like above. Alternatively, if you don't mind other metamorphic constants being on, you can add `CockroachBinary: registry.RuntimeAssertionsCockroach,` to the test spec. This will force it to always use the binary with crdb_test.

Sidenote: the benchmark test spec will force this test to always use the non crdb_test binary, unless you add the test spec mentioned above
</blockquote></details>

Thanks for chiming in! It's really useful. I will discuss with Yev to see if there is a better way of handling this. 

I think I vaguely remember Jay mentioning to me that roachtests are intended to test how production code works, so it might not have a flag like `buildutil.CrdbTestBuild` but I'm not too sure.

**Comment by wenyihu6** [comment]
*[`pkg/cmd/roachtest/tests/cdc.go` line 1125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjPJDqF-CCbLJpV59hH:-Njh950M9pVs5Khwf13w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/cmd/roachtest/tests/cdc.go#L1125)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

let's add a todo to support other formats (cause otherwise, it's not much of metamorphic test).
</blockquote></details>

Done.

**Comment by wenyihu6** [comment]
*[`pkg/cmd/roachtest/tests/cdc_helper.go` line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOSkZyDTh68fA5t3UX:-Njh9Afy2f5oipZS6SP5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/cmd/roachtest/tests/cdc_helper.go#L122)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

we probably want to use ctx here, instead of background.
</blockquote></details>

Done.

**Comment by wenyihu6** [comment]
Friendly ping on the review for this PR

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114504)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/parquet.go` line 352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjPGoZN33w2wNQdvMXj:-NkQblxa7LoNP1sIGYfj:be17zl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/ccl/changefeedccl/parquet.go#L352)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

1. (inefficiency of multiple calls to the method) Makes sense. For different parquet files but for same table, will the extra column be placed on the same index? If yes, we could cache the index for the metadata and deserialize only once.

2. (extract to a subpackage) There is some circular dependency here since both the test functions and changefeedccl code rely on `metaSentinel, parquetCrdbEventTypeColName, parquetOptUpdatedTimestampColName, parquetOptMVCCTimestampColName, parquetOptDiffColName` and deserializeMap. We could mark export them, but I'm not sure if that is wanted.
</blockquote></details>

We could export those constants, that's probably fine...
Alternatively, as this is a test only method (used in roachtest only) -- just rename this to TestingGetEventTypeColIdx -- to make it clear it's a test only method.
___
*[`pkg/util/parquet/writer.go` line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOPH0c6bvVR2tUCzDp:-NkQ_tSu9h4_XTJRCCQk:blpjr33) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/util/parquet/writer.go#L184)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

There are two pieces of metadata needed - one is for reader utility (parquet.ReadFile) to work (set in parquet package) and another is to include the metadata info for the extra column so that we can remove this extra column (set in changefeedccl package).
</blockquote></details>

But... This writer in this package is constructed from CDC package -- why can't that package instruct
this package to add metadata?
It looks like, instead of adding the same "includeParquetReaderMetadata" a better option is to 
add an explicit Option (WithParquetReaderMetadata) to enable this functionality.

**Comment by wenyihu6** [comment]
*[`pkg/ccl/changefeedccl/parquet.go` line 352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjPGoZN33w2wNQdvMXj:-NkTI1YJ2HXBrThGGaUE:b2azh52) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/ccl/changefeedccl/parquet.go#L352)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

We could export those constants, that's probably fine...
Alternatively, as this is a test only method (used in roachtest only) -- just rename this to TestingGetEventTypeColIdx -- to make it clear it's a test only method.
</blockquote></details>

Went with the easier way - renamed it to `TestingGetEventTypeColIdx`.

**Review by wenyihu6 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114504)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava and @miretskiy)

___
*[`pkg/util/parquet/writer.go` line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114504#-NjOPH0c6bvVR2tUCzDp:-NkTI8zs5of3NcFlVCNO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ccf243b0eeaa1bc53c4030e79a4240925b6ab77/pkg/util/parquet/writer.go#L184)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

But... This writer in this package is constructed from CDC package -- why can't that package instruct
this package to add metadata?
It looks like, instead of adding the same "includeParquetReaderMetadata" a better option is to 
add an explicit Option (WithParquetReaderMetadata) to enable this functionality.
</blockquote></details>

Done.

**Review by jayshrivastava — APPROVED** [review]
Very nice work!

**Comment by wenyihu6** [comment]
Thanks for the review!!

bors r=jayshrivastava

File: pkg/ccl/changefeedccl/parquet.go [hunk 1]
```diff
@@ -115,6 +118,7 @@ func newParquetWriterFromRow(
 		if opts, err = addParquetTestMetadata(row, encodingOpts, opts); err != nil {
 			return nil, err
 		}
+		opts = append(opts, parquet.WithMetadata(parquet.MakeReaderMetadata(schemaDef)))
```
Comment (wenyihu6) [inline pkg/ccl/changefeedccl/parquet.go:121]: @jayshrivastava do you know if this change will cause any existing tests to fail - it adds an additional metadata column which indicates the type of event for any buildutil.CrdbTestBuild builds

**Comment by wenyihu6** [comment]
bors r=jayshrivastava

**Comment by wenyihu6** [comment]
bors r=jayshrivastava

**Comment by wenyihu6** [comment]
bors r=jayshrivastava

**Comment by wenyihu6** [comment]
bors r=jayshrivastava

**Pull request merged**